mlir/lib/Dialect/Presburger/PresburgerDialect.cpp

```diff
@@ -28,29 +28,58 @@ PresburgerDialect::PresburgerDialect(mlir::MLIRContext *context)
 
 Attribute PresburgerDialect::parseAttribute(DialectAsmParser &parser,
                                             Type attrT) const {
+
   auto callback = [&parser](SMLoc loc, const Twine &msg) {
     return parser.emitError(loc, msg);
   };
 
-  PresburgerSetParser p(parser.getFullSymbolSpec(), callback);
-  PresburgerSet set;
+  Parser p(parser.getFullSymbolSpec(), callback);
 
-  if (failed(p.parsePresburgerSet(set))) {
-    return Attribute();
-  }
+  // Parse the kind keyword first.
+  std::unique_ptr<VariableExpr> attrKind;
+  if (failed(p.parseVariable(attrKind)))
```

Comment:
I do not like this hack, especially how it uses parseVariable, even though it parsers a special kind of keyword: `set` or `pwExpr`. I think this could be integrated deeper into the parser.